File: src/armor_detector_py/armor_detector_py/detector_node.py
#!/usr/bin/env python3
import rclpy
from rclpy.node import Node
from sensor_msgs.msg import Image
from cv_bridge import CvBridge
import cv2
import numpy as np
import torch  # <--- 加上这一行！
from ultralytics import YOLO
from aim_interfaces.msg import AimInfo 
import os

class ArmorDetectorNode(Node):
    def __init__(self):
        super().__init__('armor_detector_node')
        
        # 1. 订阅与发布
        self.sub = self.create_subscription(Image, '/sensor_img', self.img_callback, 10)
        self.pub = self.create_publisher(AimInfo, '/aim_target', 10)
        self.bridge = CvBridge()
        
        # 2. 加载模型
        # 获取当前脚本所在目录，确保能找到 best.pt
        curr_dir = os.path.dirname(os.path.abspath(__file__))
        # 这一行必须是你的绝对路径！不能有 os.path.dirname 那些东西了
        model_path = '/home/jjj/Desktop/ros2_ws/src/armor_detector_py/armor_detector_py/best.pt'
        self.get_logger().info(f'正在加载模型: {model_path} ...')
        try:
            self.model = YOLO(model_path)
            self.get_logger().info('✅ YOLOv8 模型加载成功！(PyTorch Backend)')
        except Exception as e:
            self.get_logger().error(f'❌ 模型加载失败: {e}')

        # 3. 初始化相机内参 (题目给定)
        self.K = np.array([[1462.3697, 0, 398.59394],
                           [0, 1469.68385, 110.68997],
                           [0, 0, 1]], dtype=np.float64)
        self.D = np.array([0.003518, -0.311778, -0.016581, 0.023682, 0.0000], dtype=np.float64)

        # 4. 定义装甲板 3D 坐标 (宽16cm, 高8cm)
        # 顺序: 左上, 右上, 右下, 左下
        hw, hh = 0.08, 0.04
        self.obj_pts = np.array([
            [-hw, -hh, 0], 
            [hw, -hh, 0],  
            [hw, hh, 0],   
            [-hw, hh, 0]   
        ], dtype=np.float64)

        # 5. 坐标转换所需的旋转矩阵 (预先计算好)
        r_pitch = 60 * np.pi / 180.0
        r_yaw = 20 * np.pi / 180.0
        # Ry (绕Y轴旋转)
        Ry = np.array([[np.cos(r_pitch), 0, np.sin(r_pitch)],
                       [0, 1, 0],
                       [-np.sin(r_pitch), 0, np.cos(r_pitch)]])
        # Rz (绕Z轴旋转)
        Rz = np.array([[np.cos(r_yaw), -np.sin(r_yaw), 0],
                       [np.sin(r_yaw), np.cos(r_yaw), 0],
                       [0, 0, 1]])
        # 复合旋转 R = Rz * Ry
        self.R_robot = Rz @ Ry
        # 平移 T
        self.T_robot = np.array([[0.08], [0.0], [0.05]])

    def img_callback(self, msg):
        try:
            # 将 ROS 图片转换为 OpenCV 格式
            frame = self.bridge.imgmsg_to_cv2(msg, "bgr8")
        except Exception as e:
            self.get_logger().error(f'CV Bridge 转换错误: {e}')
            return

        # --- YOLO 推理 ---
        # verbose=False 防止终端刷屏
        results = self.model(frame, verbose=False)
        
        # 遍历推理结果
        for r in results:
            boxes = r.boxes
            if len(boxes) > 0:
                # 策略: 只取置信度最高的一个目标
                # 找到 conf 最大的索引
                max_conf_idx = torch.argmax(boxes.conf).item() if hasattr(boxes.conf, 'cpu') else np.argmax(boxes.conf.cpu().numpy())
                
                best_box = boxes[max_conf_idx]
                conf = float(best_box.conf)
                cls_id = int(best_box.cls)

                if conf > 0.5: # 阈值过滤
                    # 获取方框坐标 xyxy
                    x1, y1, x2, y2 = best_box.xyxy[0].cpu().numpy()
                    
                    # 画框 (可视化)
                    cv2.rectangle(frame, (int(x1), int(y1)), (int(x2), int(y2)), (0, 255, 0), 2)
                    
                    # --- PnP 解算 ---
                    # 因为是检测模型，我们用方框的四个角近似装甲板的四个角
                    # 顺序必须对应: 左上, 右上, 右下, 左下
                    img_pts = np.array([
                        [x1, y1], 
                        [x2, y1], 
                        [x2, y2], 
                        [x1, y2]
                    ], dtype=np.float64)
                    
                    # 核心解算函数
                    success, rvec, tvec = cv2.solvePnP(self.obj_pts, img_pts, self.K, self.D)
                    
                    if success:
                        # --- 坐标转换 (相机系 -> 机器人系) ---
                        # P_robot = R * P_cam + T
                        P_cam = tvec.reshape(3, 1)
                        P_robot = self.R_robot @ P_cam + self.T_robot
                        
                        # 提取坐标 (米 -> 毫米)
                        x_mm = int(P_robot[0, 0] * 1000)
                        y_mm = int(P_robot[1, 0] * 1000)
                        z_mm = int(P_robot[2, 0] * 1000)
                        
                        # --- 发布消息 ---
                        msg_out = AimInfo()
                        msg_out.type = 7 # 题目要求固定输出 7
                        msg_out.coordinate = [x_mm, y_mm, z_mm]
                        self.pub.publish(msg_out)
                        
                        # 在图上显示距离
                        label = f"Target: {cls_id} Dist: {x_mm}mm"
                        cv2.putText(frame, label, (int(x1), int(y1)-10), 
                                    cv2.FONT_HERSHEY_SIMPLEX, 0.7, (0, 255, 255), 2)
                        
                        self.get_logger().info(f"发布目标: Type=7, Pos=[{x_mm}, {y_mm}, {z_mm}]")

        # 显示画面
        cv2.imshow("Python YOLO Detector", frame)
        cv2.waitKey(1)

def main(args=None):
    rclpy.init(args=args)
    node = ArmorDetectorNode()
    rclpy.spin(node)
    node.destroy_node()
    rclpy.shutdown()

if __name__ == '__main__':
    import torch # 延迟导入，防止某些环境冲突
    main()

File: src/aim_detector/src/detector_node.cpp
#include <rclcpp/rclcpp.hpp>
#include <sensor_msgs/msg/image.hpp>
#include <cv_bridge/cv_bridge.h>
#include <opencv2/opencv.hpp>
#include <aim_interfaces/msg/aim_info.hpp>
#include <vector>
#include <algorithm>
#include <cmath>
#include <string>
#include <opencv2/dnn.hpp>
#include <ament_index_cpp/get_package_share_directory.hpp>

using namespace std::chrono_literals;

class ArmorDetectorNode : public rclcpp::Node
{
public:
    ArmorDetectorNode() : Node("armor_detector_node")
    {
        subscription_ = this->create_subscription<sensor_msgs::msg::Image>(
            "/sensor_img", 10,
            std::bind(&ArmorDetectorNode::image_callback, this, std::placeholders::_1));

        publisher_ = this->create_publisher<aim_interfaces::msg::AimInfo>("/aim_target", 10);

        // -------------------------------------------------------------
        // 模型路径 (请确认文件已覆盖为“普通检测版”的 onnx)
        std::string model_path = "/home/jjj/Desktop/ros2_ws/src/aim_detector/model/best.onnx";
        // -------------------------------------------------------------
        
        // 必须和你导出时的尺寸一致 (416)
        input_size_ = cv::Size(416, 416); 

        try {
            net_ = cv::dnn::readNetFromONNX(model_path);
            net_.setPreferableBackend(cv::dnn::DNN_BACKEND_CUDA);
            net_.setPreferableTarget(cv::dnn::DNN_TARGET_CUDA);
            RCLCPP_INFO(this->get_logger(), "YOLO-Detect (普通版) 模型加载成功!");
        } catch (const cv::Exception& e) {
            RCLCPP_ERROR(this->get_logger(), "模型加载失败: %s", e.what());
        }

        camera_matrix_ = (cv::Mat_<double>(3, 3) << 
            1462.3697, 0, 398.59394,
            0, 1469.68385, 110.68997,
            0, 0, 1);
        dist_coeffs_ = (cv::Mat_<double>(1, 5) << 
            0.003518, -0.311778, -0.016581, 0.023682, 0.0000);

        // 装甲板 3D 坐标
        double hw = 0.16 / 2.0; 
        double hh = 0.08 / 2.0; 
        object_points_ = {
            cv::Point3f(-hw, -hh, 0), cv::Point3f(hw, -hh, 0),  
            cv::Point3f(hw, hh, 0),   cv::Point3f(-hw, hh, 0)   
        };
    }

private:
    void image_callback(const sensor_msgs::msg::Image::SharedPtr msg)
    {
        cv_bridge::CvImagePtr cv_ptr;
        try { cv_ptr = cv_bridge::toCvCopy(msg, sensor_msgs::image_encodings::BGR8); }
        catch (cv_bridge::Exception& e) { return; }

        cv::Mat frame = cv_ptr->image;
        if(frame.empty()) return;

        std::vector<cv::Point2f> image_points;
        int detected_type = 0; // 识别到的装甲板 ID

        // 调用普通检测函数
        bool found = detect_armor_yolo(frame, image_points, detected_type);

        if (found) {
            cv::Mat rvec, tvec;
            // 使用方框的四角进行 PnP
            bool success = cv::solvePnP(object_points_, image_points, camera_matrix_, dist_coeffs_, rvec, tvec);

            if (success) {
                cv::Mat p_robot = transform_camera_to_robot(tvec);
                int x_mm = (int)(p_robot.at<double>(0, 0) * 1000);
                int y_mm = (int)(p_robot.at<double>(1, 0) * 1000);
                int z_mm = (int)(p_robot.at<double>(2, 0) * 1000);

                auto result_msg = aim_interfaces::msg::AimInfo();
                result_msg.coordinate = { (int16_t)x_mm, (int16_t)y_mm, (int16_t)z_mm };
                result_msg.type = 7; 
                publisher_->publish(result_msg);

                // 画框
                for(int i=0; i<4; i++) cv::line(frame, image_points[i], image_points[(i+1)%4], cv::Scalar(0,255,0), 2);
                
                std::string text = "ID:" + std::to_string(detected_type) + " Dist:" + std::to_string(x_mm);
                cv::putText(frame, text, image_points[0], cv::FONT_HERSHEY_SIMPLEX, 0.8, cv::Scalar(0, 255, 255), 2);
            }
        }
        cv::imshow("YOLO-Detect", frame);
        cv::waitKey(1);
    }

    // --- 核心：普通检测模型解析 ---
    bool detect_armor_yolo(const cv::Mat& frame, std::vector<cv::Point2f>& detected_points, int& type_id) {
        cv::Mat blob;
        cv::dnn::blobFromImage(frame, blob, 1.0/255.0, input_size_, cv::Scalar(), true, false);
        net_.setInput(blob);
        std::vector<cv::Mat> outputs;
        net_.forward(outputs, net_.getUnconnectedOutLayersNames());

        cv::Mat output_data = outputs[0];
        
        // 处理维度问题 (适应 OpenCV 4.5.4)
        int dimensions = output_data.size[1]; 
        int rows = output_data.size[2];       
        
        if (dimensions > rows) {
            output_data = output_data.reshape(1, dimensions);
            cv::transpose(output_data, output_data);
            // 交换后：rows=8400, dimensions=4+类别数
            rows = output_data.size[0];
            dimensions = output_data.size[1];
        } else {
            // 如果原本就是 [8400, 19]
             rows = output_data.size[0];
             dimensions = output_data.size[1];
        }
        
        float* data = (float*)output_data.data;
        float x_factor = (float)frame.cols / input_size_.width;
        float y_factor = (float)frame.rows / input_size_.height;

        std::vector<float> confidences;
        std::vector<cv::Rect> boxes;
        std::vector<int> class_ids;

        // 遍历 8400 个框
        // 格式: [x, y, w, h, score1, score2, ... score14]
        for (int i = 0; i < rows; ++i) {
            float* classes_scores = data + 4; // 第4位之后全是类别分数
            
            // 找出分数最高的那个类别
            // 假设你有 14 个类别，dimensions 应该是 4 + 14 = 18
            int num_classes = dimensions - 4; 
            cv::Mat scores(1, num_classes, CV_32F, classes_scores);
            cv::Point class_id_point;
            double max_class_score;
            cv::minMaxLoc(scores, 0, &max_class_score, 0, &class_id_point);

            if (max_class_score > 0.5) { 
                float x = data[0];
                float y = data[1];
                float w = data[2];
                float h = data[3];

                int left = int((x - 0.5 * w) * x_factor);
                int top = int((y - 0.5 * h) * y_factor);
                int width = int(w * x_factor);
                int height = int(h * y_factor);

                boxes.push_back(cv::Rect(left, top, width, height));
                confidences.push_back((float)max_class_score);
                class_ids.push_back(class_id_point.x);
            }
            data += dimensions;
        }

        // NMS
        std::vector<int> indices;
        cv::dnn::NMSBoxes(boxes, confidences, 0.5, 0.45, indices);

        if (indices.size() > 0) {
            int idx = indices[0];
            type_id = class_ids[idx];
            cv::Rect best_box = boxes[idx];

            // 提取四角 (近似 PnP)
            detected_points.clear();
            detected_points.push_back(cv::Point2f(best_box.x, best_box.y)); // 左上
            detected_points.push_back(cv::Point2f(best_box.x + best_box.width, best_box.y)); // 右上
            detected_points.push_back(cv::Point2f(best_box.x + best_box.width, best_box.y + best_box.height)); // 右下
            detected_points.push_back(cv::Point2f(best_box.x, best_box.y + best_box.height)); // 左下
            
            return true;
        }
        return false;
    }

    cv::Mat transform_camera_to_robot(const cv::Mat& t_camera) {
        double r_p = 60 * M_PI / 180.0;
        double r_y = 20 * M_PI / 180.0;
        cv::Mat Ry = (cv::Mat_<double>(3, 3) << cos(r_p), 0, sin(r_p), 0, 1, 0, -sin(r_p), 0, cos(r_p));
        cv::Mat Rz = (cv::Mat_<double>(3, 3) << cos(r_y), -sin(r_y), 0, sin(r_y), cos(r_y), 0, 0, 0, 1);
        return Rz * Ry * t_camera + (cv::Mat_<double>(3, 1) << 0.08, 0.0, 0.05);
    }

    rclcpp::Subscription<sensor_msgs::msg::Image>::SharedPtr subscription_;
    rclcpp::Publisher<aim_interfaces::msg::AimInfo>::SharedPtr publisher_;
    cv::Mat camera_matrix_, dist_coeffs_;
    std::vector<cv::Point3f> object_points_;
    cv::dnn::Net net_;
    cv::Size input_size_;
};

int main(int argc, char * argv[]) {
    rclcpp::init(argc, argv);
    rclcpp::spin(std::make_shared<ArmorDetectorNode>());
    rclcpp::shutdown();
    return 0;
}